Superadmin Implementation

BrandonCookeDev opened the pull request:
This PR implements:
- Superadmin vs Region Admin privileges
- New UI on the Admin Functions that shows every user some information and allows them to change password

BrandonCookeDev commented:
I still need to implement a unit test or two before we PR this

BrandonCookeDev commented:
Tests written and this PR is ready to be reviewed

jschnei commented on model.py:
```diff
@@ -348,7 +349,8 @@ class User(orm.Document):
               ('hashed_password', orm.StringField(required=True)),
-              ('admin_regions', orm.ListField(orm.StringField()))]
+              ('admin_regions', orm.ListField(orm.StringField())),
+              ('admin_level', orm.StringField(required=True, default='REGION'))]
```
since these guys should come from an Enum, probably it is a good idea to add a choices validator so the ORM will automatically check for this. (take a look at how tournament source types work on line 129).

BrandonCookeDev replied:
So I've gone with the String Only method, but I'm not really sure how to validate on the ORM. can you point me in the right direction?

jschnei replied:
so, take a look at line 129 to see how it's done for tournament source_types (challonge, tio, smashgg).

basically, you should do something like
```
              ('admin_level', orm.StringField(
                  required=True,
                  validators=[orm.validate_choices(ADMIN_LEVEL_CHOICES)]))
```

and define a list/tuple ADMIN_LEVEL_CHOICES somewhere.

jschnei commented on model.py:
```diff
@@ -371,3 +373,8 @@ class Session(orm.Document):
               ('user_id', orm.StringField(required=True))]
+
+
+class AdminLevels(Enum):
```
probably makes most sense to move this up to the top, next to the SOURCE_TYPE_CHOICES "enum".

For that matter, I'm still not sure if we gain anything by making this an explicit enum instead of just writing a tuple of choices (as in SOURCE_TYPE_CHOICES) and using the validator. It would make more sense I feel if we were mapping human-readable enum values to non-human readable int values (this is what a lot of enums are for), but since we are mapping them to string values, it feels a bit redundant (also, in server.py, it looks like we actually just use the explicit string constants).

Anyway tl;dr either way (enums/lists of strings) is okay, but if we decide to go the Enum route we should probably also convert SOURCE_TYPE_CHOICES to an Enum for consistency.

BrandonCookeDev replied:
Going with the string method so no need for this

jschnei commented on server.py:
```diff
@@ -137,6 +144,8 @@ def get_user_from_request(request, dao):
 
 
 def is_user_admin_for_region(user, region):
+    if user.admin_level == 'SUPER':
```
if we want to stick with the Enum method, these 'SUPER's and everything should be M.AdminLevels.SUPERs.

BrandonCookeDev replied:
Going with the String method. so no need for this

jschnei commented on server.py:
```diff
@@ -1176,8 +1209,8 @@ def put(self):
             return 'Permission denied', 403
-        #if not is_user_admin_for_region(user, region='*'):
-        #    return 'Permission denied', 403
+        if not dao.get_is_superadmin(user.id):
```
hmm if we have the user object, can't we just check if user.admin_level=='SUPER'? This should save a mongo lookup.

jschnei commented on dao.py:
```diff
@@ -559,7 +559,7 @@ def insert_user(self, user):
 
     # throws invalidRegionsException, which is okay, as this is only used by a
```
let's remove this comment, it's no longer true with the introduction of the admin dashboard.

jschnei commented on dao.py:
```diff
@@ -645,6 +665,17 @@ def check_creds_and_get_session_id_or_none(self, username, password):
         else:
             return None
 
+    def check_creds(self, username, password):
```
we should try not to repeat code (this is basically part of check_creds_and_get_session_id_or_none). maybe let's make `check_creds` return a user object, and refactor `check_creds_and_get_session_id_or_none` to call `check_creds` first?

(this could also be postponed to a later PR, but we should make a note).

BrandonCookeDev replied:
Oh looking a little closer it might be because it finds no user in the db and then returns None...

jschnei replied:
i'm confused, it looks there are now 3 functions which repeat a lot of code?:

1. `check_creds(username, password)` which returns whether (username, password) are valid credentials.
2. `check_creds_and_get_session_id_or_none(username, password)`, which checks if (username, password) are valid credentials, and tries to return the current session
3. `get_and_verify_user_by_username`, which seems to be a duplicate of `get_user_by_username`.

Right now 1 and 2 duplicate a lot of the same code. I recommend removing `get_and_verify_user_by_username` and refactoring 1 and 2 so that check_creds returns a user object (not just a bool), and so that get_session_id calls check_creds to get the user object.

jschnei commented on dao.py:
```diff
@@ -619,6 +620,25 @@ def get_user_by_session_id_or_none(self, session_id):
+        else:
+            return False
+
+    def set_user_superadmin(self, user_id, admin_level):
```
this can be used for any admin_level, so let's call it something like `set_user_admin_level` or `update_user_admin_level`? (it might also make more sense to just make a general `update_user` function in the same vein as `update_player` and `update_tournament`).

jschnei commented on server.py:
```diff
@@ -1196,9 +1229,14 @@ def put(self):
             uregions = args['new_user_regions']
 
+            perm = None
+            if uperm == 'Super Admin': perm = M.AdminLevels.SUPER
```
for consistency, let's make the values of the permissions on the frontend match the string values of the enums on the backends (i.e. 'SUPER' and 'REGION')

jschnei commented on webapp/app/tools/admin_functions/views/admin_functions.html:
```diff
@@ -1,65 +1,106 @@
+
+                        <h4>Permission Level:</h4>
+                        <select ng-model="postParams.new_user_permissions" id="permissionLevelSelect">
+                            <option>Region Admin</option>
```
can we put value='REGION' and value='SUPER' here to pass 'REGION' and 'SUPER' to the backend? (not a big deal, but it's nice when values are consistent between frontend and backend)

jschnei commented on dao.py:
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
+            return False
+
+    def set_user_admin_level(self, user_id, admin_level):
+        if type(admin_level) is not M.AdminLevels:
```
I think M.AdminLevels does not exist anymore, so this should error (is it being tested)?
In any case, the ORM should take care of this check.

jschnei commented on dao.py:
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
         pass
 
+    def get_is_superadmin(self, user_id):
+        user = None
```
I think it is probably better to use `get_user_by_id_or_none` here instead of rewriting the Mongo queries (so something like `user = get_user_by_id_or_none(user_id)`.

Actually, do we use this function (get_is_super_admin) anywhere else in the code? If not, I think it's fine to remove this function (it is easy enough to just check whether user.admin_level=='SUPER' whenever we have a user object).

jschnei commented on dao.py:
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
+        else:
+            return False
+
+    def set_user_admin_level(self, user_id, admin_level):
```
do we actually use this function anywhere else in the code yet? if not, I think it would be fine to just replace it with a simple `update_user` function, in the same style as `update_player` and `update_region`.

jschnei commented on dao.py:
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
+    def set_user_admin_level(self, user_id, admin_level):
+        if type(admin_level) is not M.AdminLevels:
+            raise Exception('Submitted admin level is not of correct type')
+        if self.users_col.find_one({'_id': user_id}):
```
right now this way of updating bypasses the ORM completely and won't trigger any of the ORMs validators. the correct way to do this sort of thing is 1. read the user data into a User object (as is done in get_user_by_id_or_none), 2. update whatever properties of the User object, 3. then write the User object back into the database using `self.users_col.update({'_id': user.id}, user.dump(context='db'))`.

BrandonCookeDev commented:
Jon, I made the changes. Notably I deleted the third check_creds thing I had written. I left it in to try to fix whatever bug was in it, but now I have no time.
Also I rewrote the set_admin_level function, and commented it out. Since it isn't implemented yet, and we may want it in the future, I felt this was the best course of action

jschnei commented:
I fixed a bug with creating new users; I think it works now.

Also, while fixing this I noticed some of the frontend code is probably less angular-y than it should be; there's a lot of places where you're manipulating the DOM directly (I think the angular philosophy is that this should basically always be done through angular directives). Fixing this can probably wait for a later PR though; there is a bunch of frontend refactoring that still needs to be done.
